### ISSUE TITLE
Pristine-1.4: During IPCM killing sometimes ipc process hangs

### ISSUE THREAD

**Issue by kewinrausch** [IRATI/stack#873]
When I terminate the ipcm with pkill -SIGINT, sometimes instead of exit gracefully an ipc process (always the normal one, not the shim) does not get killed and assumes a _defunct_ state.

Log from kernel:
```
[ 5851.770811] rina-kipcm(ERR): Could not find IPC Process with id 2
[ 5856.778016] rina-rnl(WARN): IPC Manager process has been destroyed
```

Log from IPCM:
```
10357(1455811439)#ipcp[2] (CRIT): Got signal 2
10357(1455811439)#ipcp[2].rib-daemon (ERR): Problems reading management SDU: Unknown error
10357(1455811439)#ipcp[2].rib-daemon (ERR): Problems reading management SDU: the IPCP or the needed parts of the ipcp do not exist
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/queued_pdus
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/drop_pdus
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/err_pdus
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/rx_pdus
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/tx_pdus
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/rx_bytes
10357(1455811439)#ipcp[2].utils (ERR): Problems opening file: /sys/rina/ipcps/2/rmt/n1_ports/1/tx_bytes
10352(1455811440)#ipcm (INFO)[destroy_ipcp]: IPC process destroyed [id = 2]
```

ps -A trace:
```
root@s1:~# ps -A
  PID TTY          TIME CMD
    1 ?        00:00:03 systemd
    2 ?        00:00:00 kthreadd
    3 ?        00:00:00 ksoftirqd/0
    5 ?        00:00:00 kworker/0:0H
    7 ?        00:00:48 rcu_sched
    8 ?        00:00:00 rcu_bh
    9 ?        00:00:00 migration/0
   10 ?        00:00:00 watchdog/0
   11 ?        00:00:00 watchdog/1
   12 ?        00:00:00 migration/1
   13 ?        00:00:38 ksoftirqd/1
   15 ?        00:00:00 kworker/1:0H
   16 ?        00:00:00 khelper
   17 ?        00:00:00 kdevtmpfs
   18 ?        00:00:00 netns
   19 ?        00:00:00 perf
   20 ?        00:00:00 khungtaskd
   21 ?        00:00:00 writeback
   23 ?        00:00:00 ksmd
   24 ?        00:00:00 khugepaged
   25 ?        00:00:00 crypto
   26 ?        00:00:00 kintegrityd
   27 ?        00:00:00 bioset
   28 ?        00:00:00 kblockd
   29 ?        00:00:00 devfreq_wq
   31 ?        00:00:00 kswapd0
   32 ?        00:00:00 fsnotify_mark
   39 ?        00:00:00 kthrotld
   40 ?        00:00:00 ipv6_addrconf
   42 ?        00:00:00 rnl-wq
   43 ?        00:00:00 flowdelq
   44 ?        00:00:00 deferwq
   80 ?        00:00:00 acpi_thermal_pm
   81 ?        00:00:00 firewire
   82 ?        00:00:00 firewire_ohci
   83 ?        00:00:00 ata_sff
   84 ?        00:00:00 irq/18-mmc0
   85 ?        00:00:06 kworker/0:3
   86 ?        00:00:00 scsi_eh_0
   87 ?        00:00:00 scsi_tmf_0
   88 ?        00:00:00 scsi_eh_1
   89 ?        00:00:00 scsi_tmf_1
   90 ?        00:00:00 scsi_eh_2
   91 ?        00:00:00 scsi_tmf_2
   92 ?        00:00:00 scsi_eh_3
   93 ?        00:00:00 scsi_tmf_3
   94 ?        00:00:00 scsi_eh_4
   95 ?        00:00:00 scsi_tmf_4
   96 ?        00:00:00 scsi_eh_5
   97 ?        00:00:00 scsi_tmf_5
  109 ?        00:00:00 kworker/1:1H
  110 ?        00:00:00 kworker/0:1H
  132 ?        00:00:00 jbd2/sda1-8
  133 ?        00:00:00 ext4-rsv-conver
  164 ?        00:00:00 kauditd
  174 ?        00:00:00 systemd-journal
  180 ?        00:00:00 systemd-udevd
  408 ?        00:00:00 pccardd
  413 ?        00:00:00 kpsmoused
  416 ?        00:00:00 rpcbind
  429 ?        00:00:00 cfg80211
  445 ?        00:00:00 rpc.statd
  447 ?        00:00:28 irq/31-iwlwifi
  455 ?        00:00:00 rpciod
  458 ?        00:00:00 iwlwifi
  466 ?        00:00:00 nfsiod
  480 ?        00:00:00 rpc.idmapd
  484 ?        00:00:00 cron
  486 ?        00:00:02 NetworkManager
  490 ?        00:00:00 sshd
  491 ?        00:00:00 atd
  492 ?        00:00:00 ModemManager
  494 ?        00:00:00 systemd-logind
  499 ?        00:00:00 avahi-daemon
  500 ?        00:00:00 dbus-daemon
  509 ?        00:00:00 ttm_swap
  515 ?        00:00:00 avahi-daemon
  529 ?        00:00:00 minissdpd
  560 ?        00:00:00 sshd
  562 pts/2    00:00:00 bash
  578 ?        00:00:00 rsyslogd
  582 ?        00:00:00 acpid
  748 ?        00:00:00 kworker/1:1
  754 pts/2    00:00:00 ps
  786 ?        00:00:00 polkitd
  789 tty1     00:00:00 agetty
  792 ?        00:00:00 lightdm
  803 tty7     00:00:20 Xorg
  814 ?        00:00:00 exim4
  822 ?        00:00:00 wpa_supplicant
  834 ?        00:00:00 systemd
  835 ?        00:00:00 (sd-pam)
  842 ?        00:00:00 dbus-launch
  843 ?        00:00:00 dbus-daemon
  845 ?        00:00:00 at-spi-bus-laun
  849 ?        00:00:00 dbus-daemon
  852 ?        00:00:00 at-spi2-registr
  858 ?        00:00:00 lightdm
  863 ?        00:00:00 systemd
  864 ?        00:00:00 (sd-pam)
  868 ?        00:00:00 gnome-keyring-d
  870 ?        00:00:00 lxsession
  891 ?        00:00:00 ssh-agent
  894 ?        00:00:00 dbus-launch
  895 ?        00:00:00 dbus-daemon
  901 ?        00:00:00 gvfsd
  905 ?        00:00:00 gvfsd-fuse
  915 ?        00:00:00 openbox
  919 ?        00:00:00 lxpolkit
  921 ?        00:00:04 lxpanel
  922 ?        00:00:00 pcmanfm
  927 ?        00:00:00 ssh-agent
  929 ?        00:00:00 applet.py
  932 ?        00:00:00 notification-da
  941 ?        00:00:01 nm-applet
  943 ?        00:00:00 at-spi-bus-laun
  947 ?        00:00:00 dbus-daemon
  950 ?        00:00:00 at-spi2-registr
  955 ?        00:00:00 gvfs-udisks2-vo
  957 ?        00:00:01 udisksd
  967 ?        00:00:18 clipit
  971 ?        00:00:00 gvfs-mtp-volume
  975 ?        00:00:00 gvfs-gphoto2-vo
  979 ?        00:00:00 pulseaudio
  982 ?        00:00:00 gconfd-2
  986 ?        00:00:00 gvfs-goa-volume
  990 ?        00:00:00 gvfs-afc-volume
  997 ?        00:00:00 gnome-keyring-d
 1007 ?        00:00:00 menu-cached
 1011 ?        00:00:00 gvfsd-trash
 1023 ?        00:00:00 start-pulseaudi
 1024 ?        00:00:00 xprop
 1049 ?        00:00:00 dhclient
 1881 ?        00:00:00 sshd
 1883 pts/0    00:00:00 bash
 3749 ?        00:00:00 arp826-wq
 3750 ?        00:00:00 kworker/u5:0
 3751 ?        00:00:00 shim-eth-vlan
 4004 ?        00:00:00 sshd
 4006 pts/1    00:00:00 bash
 7956 ?        00:00:02 kworker/1:3
10352 ?        00:00:06 ipcm
10357 ?        00:00:08 ipcp <defunct>
25568 ?        00:00:00 kworker/u4:0
29480 ?        00:00:00 kworker/0:1
30545 ?        00:00:00 kworker/1:0
30887 ?        00:00:00 kworker/u4:1
32091 ?        00:00:00 kworker/0:0
32407 ?        00:00:00 kworker/u4:2
```

**Comment by kewinrausch** [comment]
I'm still proceeding investigating this issue... an additional log on this is it's stack trace which I obtained while investigating it's proc entry:

```
[...]
 1140 ?        00:00:00 gnome-pty-helpe
 1141 pts/0    00:00:00 bash
 1147 ?        00:00:00 sshd
 1149 pts/1    00:00:00 bash
 1232 ?        00:00:01 kworker/1:0
 1273 ?        00:00:00 arp826-wq
 1274 ?        00:00:00 shim-eth-vlan
 1347 ?        00:00:00 kworker/0:0
 2885 ?        00:00:00 kworker/0:2
 2928 ?        00:00:00 kworker/0:3
 3065 ?        00:00:00 kworker/u4:1
 3440 ?        00:00:00 kworker/1:2
 4387 ?        00:00:00 kworker/0:1
 4396 ?        00:00:00 kworker/u4:0
 4430 ?        00:00:00 ipcp <defunct>
 4990 ?        00:00:00 kworker/1:1
 5227 ?        00:00:00 kworker/1:3
 5246 pts/1    00:00:00 ps
root@s2:~# cat /proc/4430/stack
[<ffffffff81072bc2>] do_exit+0x782/0xad0
[<ffffffff81072f89>] do_group_exit+0x39/0xb0
[<ffffffff8107e2f8>] get_signal+0x2b8/0x6a0
[<ffffffff810124f3>] do_signal+0x23/0xb30
[<ffffffff810e6773>] SyS_futex+0x73/0x170
[<ffffffff81013078>] do_notify_resume+0x78/0x90
[<ffffffff81590184>] int_signal+0x12/0x17
[<ffffffffffffffff>] 0xffffffffffffffff
```

**Comment by vmaffione** [comment]
@kewinrausch is this bug still there?

**Comment by kewinrausch** [comment]
Yes, it is. Just got it now:
```
[...]
 3679 ?        00:00:00 rc.progagent
 3680 ?        00:00:00 program-agent
 3707 ?        00:00:00 linktest
 3716 tty1     00:00:00 getty
 3717 tty2     00:00:00 getty
 3718 tty3     00:00:00 getty
 3719 tty4     00:00:00 getty
 3720 tty5     00:00:00 getty
 3721 tty6     00:00:00 getty
 3722 ?        00:00:00 kworker/u50:1
 3913 ?        00:00:00 kworker/0:0
 4287 ?        00:00:00 arp826-wq
 4288 ?        00:00:00 kworker/u51:0
 4289 ?        00:00:00 kworker/u52:0
 4290 ?        00:00:00 kworker/u53:0
 4291 ?        00:00:00 shim-eth-vlan
 4299 ?        00:00:02 ipcp <defunct>
 4300 ?        00:00:00 udevd
 6299 ?        00:00:00 ipcm
 6304 ?        00:00:00 ipcp
 6305 ?        00:00:00 udevd
 6421 ?        00:00:00 sshd
 6425 ?        00:00:00 sshd
 6426 pts/0    00:00:00 bash
 6491 pts/0    00:00:00 ps
```

**Comment by bernatgaston** [comment]
Hi, There is a segfault when closing the IPCP. I think that this is the reason why it hangs. This segfault is in the timer task, and has been introduced in the last weeks/months. I am addressing it

**Comment by edugrasa** [comment]
Fixed by #911
